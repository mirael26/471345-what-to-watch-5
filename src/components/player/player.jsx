import React, {PureComponent, createRef} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {fetchCurrentFilm} from "../../store/api-actions";
import {ActionCreator} from "../../store/action";
import {debounce} from "../../utils";

import TimeBar from "../time-bar/time-bar";

const DEFAULT_RUN_TIME = 5400;

class Player extends PureComponent {
  constructor(props) {
    super(props);

    this._videoRef = createRef();
    this._fullscreenButtonRef = createRef();
    this._duration = DEFAULT_RUN_TIME;

    this.onFullscreenButtonClick = this.onFullscreenButtonClick.bind(this);
    this.changeTime = this.changeTime.bind(this);
    this.sendCurrentTime = this.sendCurrentTime.bind(this);
  }

  componentDidMount() {
    const {loadCurrentFilm} = this.props;
    loadCurrentFilm(this.getFilmId());

    const video = this._videoRef.current;
    video.addEventListener(`durationchange`, () => this.setDuration());
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.id !== prevProps.match.params.id) {
      this.props.loadCurrentFilm(this.getFilmId());
    }

    if (prevProps.isPlaying !== this.props.isPlaying) {
      this.changePlayMode();
    }
  }

  getFilmId() {
    const {match} = this.props;
    const {params: {id}} = match;
    return parseInt(id, 10);
  }

  changePlayMode() {
    const {isPlaying} = this.props;
    const video = this._videoRef.current;
    if (isPlaying) {
      video.play();
    } else {
      video.pause();
    }
  }

  onFullscreenButtonClick() {
    const video = this._videoRef.current;
    if (video.requestFullscreen) {
      video.requestFullscreen();
    } else if (video.webkitRequestFullscreen) {
      video.webkitRequestFullscreen();
    } else if (video.msRequestFullscreen) {
      video.msRequestFullscreen();
    }
  }

  sendCurrentTime() {
    const currentTime = this._videoRef.current.currentTime;
    const {setCurrentTime} = this.props;
    setCurrentTime(currentTime);
  }

  setDuration() {
    this._duration = this._videoRef.current.duration;
  }

  changeTime(newTime) {
    this._videoRef.current.currentTime = newTime;
  }

  render() {
    const {
      currentFilm,
      onExitButtonClick,
      isPlaying,
      onPlayButtonClick,
    } = this.props;

    return (
      <div className="player">
        <video
          src={currentFilm.videoSrc}
          className="player__video"
          ref={this._videoRef}
          onTimeUpdate={debounce(this.sendCurrentTime, 1000)}></video>

        <button type="button" className="player__exit" onClick={onExitButtonClick}>Exit</button>

        <div className="player__controls">
          <TimeBar
            runTime={this._duration}
            changeTime={this.changeTime}
            changePlayMode={onPlayButtonClick}
            isPlaying={isPlaying}/>

          <div className="player__controls-row">
            {isPlaying
              ? <button type="button" className="player__play" onClick={onPlayButtonClick}>
                <svg viewBox="0 0 14 21" width="14" height="21">
                  <use xlinkHref="#pause"></use>
                </svg>
                <span>Pause</span>
              </button>
              : <button type="button" className="player__play" onClick={onPlayButtonClick}>
                <svg viewBox="0 0 19 19" width="19" height="19">
                  <use xlinkHref="#play-s"></use>
                </svg>
                <span>Play</span>
              </button>}

            <div className="player__name">Transpotting</div>

            <button
              type="button"
              className="player__full-screen"
              ref={this._fullscreenButtonRef}
              onClick={this.onFullscreenButtonClick}>
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"></use>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

Player.propTypes = {
  currentFilm: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    genre: PropTypes.string,
    releaseDate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    description: PropTypes.string,
    picture: PropTypes.string,
    poster: PropTypes.string,
    backgroundPicture: PropTypes.string,
    rate: PropTypes.number,
    voteCount: PropTypes.number,
    director: PropTypes.string,
    starring: PropTypes.arrayOf(PropTypes.string),
    runTime: PropTypes.number,
    previewVideoSrc: PropTypes.string,
    videoSrc: PropTypes.string,
  }),
  match: PropTypes.object.isRequired,
  loadCurrentFilm: PropTypes.func.isRequired,
  onExitButtonClick: PropTypes.func.isRequired,
  isPlaying: PropTypes.bool.isRequired,
  onPlayButtonClick: PropTypes.func.isRequired,
  setCurrentTime: PropTypes.func.isRequired,
};

const mapStateToProps = ({DATA}) => ({
  currentFilm: DATA.currentFilm,
});

const mapDispatchToProps = (dispatch) => ({
  loadCurrentFilm(id) {
    dispatch(fetchCurrentFilm(id));
  },
  setCurrentTime(time) {
    dispatch(ActionCreator.setPlayerTime(time));
  }
});

export {Player};
export default connect(mapStateToProps, mapDispatchToProps)(Player);
