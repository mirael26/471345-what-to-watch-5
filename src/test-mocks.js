const films = [
  {
    id: 1,
    title: `The Grand Budapest Hotel`,
    genre: `Drama`,
    releaseDate: `2014`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.

    Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
    picture: `https://www.theglobeandmail.com/resizer/ZlZ4o9jXtk_x8e20WPaKO1sFD6I=/1200x0/filters:quality(80)/arc-anglerfish-tgam-prod-tgam.s3.amazonaws.com/public/CEK54CW3HZBMXKO3QNV6PHHNEE`,
    poster: `img/the-grand-budapest-hotel-poster.jpg`,
    backgroundPicture: `img/bg-the-grand-budapest-hotel.jpg`,
    backgroundColor: `ffffff`,
    rate: 8.9,
    voteCount: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    runTime: 99,
    previewVideoSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    videoSrc: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    isFavorite: true,
  },
  {
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence ,and redemption.`,
    director: `Quentin Tarantino`,
    genre: `Crime`,
    id: 2,
    title: `Pulp Fiction`,
    releaseDate: 1994,
    picture: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    backgroundPicture: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    backgroundColor: `#795433`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    rate: 1.5,
    voteCount: 1635992,
    runTime: 153,
    videoSrc: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideoSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.,ogv.360p.webm`,
    isFavorite: false,
  },
  {
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers ,fight to track down a priceless stolen diamond.`,
    director: `Guy Ritchie`,
    genre: `Comedy`,
    id: 13,
    title: `Snatch`,
    releaseDate: 2000,
    picture: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    poster: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    backgroundPicture: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    backgroundColor: `#FDFDFC`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    rate: 0.2,
    voteCount: 716577,
    runTime: 104,
    videoSrc: `http: //media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideoSrc: `https: //upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.,360p.webm`,
    isFavorite: false,
  },
];

const reviews = [
  {
    id: 1,
    text: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    rate: 8.9,
    author: `Kate Muir`,
    date: `2016-12-24`,
  },
  {
    id: 2,
    text: `Anderson's films are too precious for some, but for those of us willing to lose ourselves in them, they're a delight. "The Grand Budapest Hotel" is no different, except that he has added a hint of gravitas to the mix, improving the recipe.`,
    rate: 8.0,
    author: `Bill Goodykoontz`,
    date: `2015-11-18`,
  },
  {
    id: 3,
    text: `I didn't find it amusing, and while I can appreciate the creativity, it's an hour and 40 minutes I wish I could take back.`,
    rate: 8.0,
    author: `Amanda Greever`,
    date: `2015-11-18`,
  },
  {
    id: 4,
    text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
    rate: 7.2,
    author: `Matthew Lickona`,
    date: `2016-12-20`,
  },
];

const genres = [
  `Action`,
  `Crime`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Fantasy`,
  `Thriller`,
];

const match = {
  params: {
    id: 1,
  }
};

export {films, reviews, genres, match};
