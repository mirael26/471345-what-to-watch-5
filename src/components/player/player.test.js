import React from "react";
import renderer from "react-test-renderer";

import {films, match} from "../../test-mocks";

import {Player} from "./player";

jest.mock(`../time-bar/time-bar`, () => `TimeBar`);

it(`Should Player render correctly`, () => {
  const noop = () => {};
  const tree = renderer
    .create(
        <Player
          currentFilm={films[1]}
          match={match}
          loadCurrentFilm={noop}
          onExitButtonClick={noop}
          isPlaying={false}
          onPlayButtonClick={noop}
          setCurrentTime={noop}
        />
        , {
          createNodeMock: () => {
            return {};
          }
        }
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
