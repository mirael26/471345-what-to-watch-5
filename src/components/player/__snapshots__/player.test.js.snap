// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Player render correctly 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    muted={true}
    onTimeUpdate={[Function]}
    src="http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <TimeBar
      isPlaying={false}
      onPlayModeChange={[Function]}
      onTimeChange={[Function]}
      runTime={5400}
    />
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Transpotting
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
